pkg/nodeagent/controller/certificate/reconciler.go

```diff
@@ -0,0 +1,95 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package certificate
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/spf13/afero"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"k8s.io/utils/clock"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/nodeagent"
+	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
+)
+
+// Reconciler checks when the certificate of gardener-node-agent expires and requests a new one in this case.
+// When the certificate is renewed it saves the resulting kubeconfig on the disk, cancels its context to initiate a
+// restart of gardener-node-agent.
+type Reconciler struct {
+	CancelContext context.CancelFunc
```

Comment:
Nit:

```suggestion
	Cancel context.CancelFunc
```

For consistency with the naming in https://github.com/gardener/gardener/blob/e25b71273b2094aca1df07e3c61b3bba084eef51/cmd/gardener-node-agent/app/bootstrappers/nodeagent_kubeconfig.go#L23